frontend/packages/ceph-storage-plugin/src/components/ocs-install/node-list.tsx

```diff
@@ -0,0 +1,330 @@
+    },
+    {
+      title: 'Memory',
+      props: { className: tableColumnClasses[4] },
+    },
+  ];
+};
+
+// return an empty array when there is no data
+const getRows = (nodes) => {
+  return nodes.map((node) => {
+    const roles = getNodeRoles(node).sort();
+    const obj = {
+      cells: [],
+      selected: false,
+      id: node.metadata.name,
+      metadata: _.clone(node.metadata),
+      spec: _.clone(node.spec),
+    };
+    obj.cells = [
+      {
+        title: <ResourceLink kind="Node" name={node.metadata.name} title={node.metadata.uid} />,
+      },
+      {
+        title: roles.join(', ') || '-',
+      },
+      {
+        title: `${humanizeCpuCores(_.get(node.status, 'capacity.cpu')).string || '-'} CPU`,
+      },
+      {
+        title: `${getConvertedUnits(_.get(node.status, 'allocatable.memory'), 'KiB')}`,
+      },
+    ];
+    return obj;
+  });
+};
+
+const getAllTableFilters = () => ({
+  ...tableFilters,
+});
+
+const getFilteredRows = (_filters, objects) => {
+  if (_.isEmpty(_filters)) {
+    return objects;
+  }
+
+  const allTableFilters = getAllTableFilters();
+  let filteredObjects = objects;
+  _.each(_filters, (value, name) => {
+    const filter = allTableFilters[name];
+    if (_.isFunction(filter)) {
+      filteredObjects = _.filter(filteredObjects, (o) => filter(value, o));
+    }
+  });
+
+  return filteredObjects;
+};
+
+const getPreSelectedNodes = (nodes) => {
+  nodes.forEach((node) => {
+    node.selected = _.get(node, `metadata.labels['${ocsLabel}']`) !== undefined;
+  });
+};
+
+const stateToProps = (obj, { data = [], filters = {}, staticFilters = [{}] }) => {
+  const allFilters = staticFilters ? Object.assign({}, filters, ...staticFilters) : filters;
+  const newData = getFilteredRows(allFilters, data);
+  return {
+    data: newData,
+    unfilteredData: data,
+  };
+};
+
+const CustomNodeTable: React.FC<CustomNodeTableProps> = ({ data, loaded, ocsProps }) => {
+  const columns = getColumns();
+  const [nodes, setNodes] = React.useState([]);
+  const [sortBy, setSortBy] = React.useState<ISortBy>({ index: 0, direction: 'asc' });
+  const [error, setError] = React.useState('');
+  const [inProgress, setProgress] = React.useState(false);
+  const [storageClass, setStorageClass] = React.useState('');
+  const [selectedNodesCnt, setSelectedNodesCnt] = React.useState(0);
+
+  React.useEffect(() => {
+    const selectedNode = _.filter(nodes, 'selected').length;
+    setSelectedNodesCnt(selectedNode);
+  }, [nodes]);
+
+  React.useEffect(() => {
+    const formattedNodes = getRows(data);
+    // pre-selection of nodes
+    if (loaded && !nodes.length) {
+      getPreSelectedNodes(formattedNodes);
+      setNodes(formattedNodes);
+    }
+    // for getting nodes
+    else if (formattedNodes.length) {
+      nodes.forEach((row) => {
+        if (row.selected) {
+          const index = formattedNodes.findIndex((r) => r.id === row.id);
+          formattedNodes[index].selected = true;
```

Comment:
We should look at how we're tracking selection. I don't think it needs to be this difficult.

Can we create a `Set` of selected node names or node UIDs? Then it's a simple check if the node is in the set to tell if it's selected when rendering the table rows, and you don't have to track selected node count separately in state. It's just `Set.length`

The only special case you'd have to handle is removing deleted nodes from the selected `Set` on a `Firehose` update.

Reply:
@spadgett Understood your point. The patternfly react table component expects `selected` property on each object, in order to make it selected or not[1]. As per my understanding, even if I maintain the Set for selected Nodes, I have to mark the selected property true/false of appropriate node(by finding out the index) in order to make it work.

1. https://patternfly-react.surge.sh/patternfly-4/components/table/